### DISCUSSION TITLE
Can I use environment variables in renovate.json?

### DISCUSSION THREAD

**Discussion by mulcahys** [renovatebot/renovate#9456] [Request Help]
Hi,

Can I use environment variables in renovate.json? If so, what is the syntax?

This is for use in a github repo - I want to keep the secrets out of the repo

e.g.

```
{
"hostRules": [
    {
      "hostType": "docker",
      "hostName": "reg.example.com",
      "username": USERNAME_ENV_VAR,
      "password": USERNAME_ENV_PASS
    }
  ]
}
```

Thanks

**Comment by mulcahys** [comment]
Hhhm, on reflection the question may not make sense - maybe the question is how can I securely make secrets available to renovate (in particular docker registry credentials, but maybe others down the line)

**Reply by rarkins** [reply]
Are you using the hosted app on GitHub.com, or self-hosted ?

**Reply by mulcahys** [reply]
hosted app on GitHub.com

**Reply by mulcahys** [reply]
I found https://github.com/renovatebot/renovate/issues/6718 already - that would have solved my use case but I can see why it's not suitable. Something like https://github.com/mozilla/sops support might do the trick.

Or is my only option in this scenario to use something like https://circleci.com/developer/orbs/orb/daniel-shuy/renovate (and use CircleCI to inject secrets which I can access like https://docs.renovatebot.com/docker/#self-hosted-docker-registry ? )

**Reply by rarkins** [reply]
Private or public repo? Check out the "encrypted" config option if private

**Reply by mulcahys** [reply]
private - so https://docs.renovatebot.com/configuration-options/#encrypted will work in hostRules too? I wasn't clear on that from https://docs.renovatebot.com/private-modules/

This may be a separate discussion but if I understand that page I need to use https://renovatebot.com/encrypt to create the encrypted strings? Is there a programmatic alternative (we frequently rotate secrets and would prefer to avoid introducing a manual step).

Thanks

**Reply by viceice** [reply]
1. yes, encrypted works in nearly every config object.
2. You can look at the encryption source how to automate it

```js
 $(document).ready(function(){
            const publicKey = `
                    -----BEGIN PUBLIC KEY-----
                    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0u6+kZ7yoy1IMUjfoDY+
                    WwaTtfaQyLmYM/bs/CtekvDFbSQnOHIunnLKFo8OQW/PeLBur+BIcfzS5spVpXB5
                    07P3yzf/mUYwX3sdy1Zu3JdcWKKho793niIAdKNQg48xipCniVg6J5l3WK5816KB
                    Dc2+Bjwer2z5cE9G1pUPRnK3m0uHrVsFxmMnk38RZcZnGmokoBzMjUa/2w1kCHuD
                    Eq3kdSHvLBmmo5bP9OHHV9F4KVlB8cDp3TSc74U0BVEUDe3BBf9VgXfvqhjDTRJh
                    lpC+QxgdBj958K/h8BnRB6vkW3l5OXirowyXg4ZAWQn0XJ+lby5w7yCg4HetyYH5
                    0QIDAQAB
                    -----END PUBLIC KEY-----
                    `;

            $('.generate_btn').click(function(){
                var input = $('#npmtoken').val();
                var crypt = new JSEncrypt();
                crypt.setKey(publicKey);
                const encrypted = crypt.encrypt(input);
                $('#generatedText').text(encrypted.toString('base64'));
            });
```

**Reply by mulcahys** [reply]
Thanks for the super quick responses folks - really appreciate it

**Reply by nagarajui7** [reply]
@rarkins , how can I securely make secrets available to self-hosted renovate

**Reply by viceice** [reply]
@nagarajui7 Please open a new discussion for your question.

**Reply by nagarajui7** [reply]
sure @viceice
